cirq/ops/measurement_gate.py

```diff
@@ -79,6 +79,11 @@ def with_key(self, key: str) -> 'MeasurementGate':
                                invert_mask=self.invert_mask,
                                qid_shape=self._qid_shape)
 
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]):
+        if self.key not in key_map:
+            return self
```

Comment:
I'm not entirely sure that this is the right way to implement this. 
It is kind of surprising to me that the object returns itself as a result of a `with_*` method and makes it harder to reason about code and mutability.

Reply:
I was conflicted on this as well. What threw me off is the `GateOperation.with_gate` method, which returns `self` if nothing changed...but that's the exception, and I've never actually seen that method called. The only other appearance of `with_gate` in the code is for `ParallelGateOperation`, and that version does not return self. Maybe `GateOperation.with_gate` should be changed?

I'll go ahead and convert `_with_measurement_key_mapping_` to return a new object for now.

Reply:
> Maybe `GateOperation.with_gate` should be changed?

I think so!!!

Reply:
Discussion from #3499: `GateOperation` can safely return `self` from `with`-functions, as it is immutable. I'll document this more clearly and continue returning `self` in `_with_measurement_key_mapping_`.